cmd/xmidt-agent/wrphandlers.go

```diff
@@ -0,0 +1,208 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+package main
+
+import (
+	"errors"
+
+	"github.com/xmidt-org/wrp-go/v3"
+	"github.com/xmidt-org/xmidt-agent/internal/pubsub"
+	"github.com/xmidt-org/xmidt-agent/internal/websocket"
+	"github.com/xmidt-org/xmidt-agent/internal/websocket/event"
+	"github.com/xmidt-org/xmidt-agent/internal/wrphandlers/auth"
+	"github.com/xmidt-org/xmidt-agent/internal/wrphandlers/missing"
+	"github.com/xmidt-org/xmidt-agent/internal/wrphandlers/mocktr181"
+	wshandler "github.com/xmidt-org/xmidt-agent/internal/wrphandlers/websocket"
+	"go.uber.org/fx"
+)
+
+var (
+	ErrWRPHandlerConfig = errors.New("wrphandler configuration error")
+)
+
+func provideWRPHandlers() fx.Option {
+	return fx.Options(
+		fx.Provide(
+			provideWSHandler,
+			providePubSubHandler,
+			provideMissingHandler,
+			provideAuthHandler,
+		),
+		fx.Invoke(provideWSEventorToHandlerAdapter),
+	)
+}
+
+type wsAdapterIn struct {
+	fx.In
+
+	// Configuration
+	WSConfg Websocket
+	WS      *websocket.Websocket
+
+	// wrphandlers
+	AuthHandler             *auth.Handler
+	WRPHandlerAdapterCancel event.CancelFunc
+}
+
+func provideWSEventorToHandlerAdapter(in wsAdapterIn) {
+	if in.WSConfg.Disable {
+		return
+	}
+
+	in.WS.AddMessageListener(
+		event.MsgListenerFunc(func(m wrp.Message) {
+			_ = in.AuthHandler.HandleWrp(m)
+		}),
+		&in.WRPHandlerAdapterCancel,
+	)
+}
+
+type wsHandlerIn struct {
+	fx.In
+
+	// Configuration
+	WSConfg Websocket
+
+	WS *websocket.Websocket
+}
+
+func provideWSHandler(in wsHandlerIn) (h *wshandler.Handler, err error) {
+	defer func() {
+		if err != nil {
+			err = errors.Join(ErrWRPHandlerConfig, err)
+		}
+	}()
+
+	if in.WSConfg.Disable {
+		return nil, nil
+	}
+
+	return wshandler.New(in.WS)
+}
+
+type missingIn struct {
+	fx.In
+
+	// Configuration
+	// Note, DeviceID and PartnerID is pulled from the Identity configuration
+	DeviceID wrp.DeviceID
+	WSConfg  Websocket
+
+	// wrphandlers
+	WSHandler *wshandler.Handler
+	Pubsub    *pubsub.PubSub
+}
+
+func provideMissingHandler(in missingIn) (h *missing.Handler, err error) {
+	defer func() {
+		if err != nil {
```

Comment:
Instead of doing this as a defer, why not make the conditional after the code it handles?  This brings more mental baggage for me.